horovod/common/gloo/http_store.cc

```diff
@@ -49,8 +49,9 @@ void HTTPStore::wait(const std::vector<std::string>& keys,
     const auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(
         std::chrono::steady_clock::now() - start);
     if (timeout != gloo::kNoTimeout && elapsed > timeout) {
-      GLOO_THROW_IO_EXCEPTION(GLOO_ERROR_MSG("Wait timeout for key(s): ",
-                                             ::gloo::MakeString(keys)));
+      auto timeout_seconds = std::chrono::duration_cast<std::chrono::seconds>(timeout);
```

Comment:
Nice!  Can you also add a suggestion here to try increasing `HOROVOD_GLOO_TIMEOUT_SECONDS`?

Reply:
Ideally, that should be in the exception message, but this part of the code does not know (and should not make this assumption) where the timeout value is coming from, it is just an argument to this method. The env var is defined and read in `gloo_context.cc`.

Ideally, the timeout exception should be caught there and re-thrown with an enhanced message saying where to set the timeout. For that, the exception must be of a very specific type, like gloo timeout exception for that to work.

My hope is seeing the timeout at all in the logs helps to realize that whatever timeout knob you used is the wrong one. Not ideal but better than nothing.

Any better idea?

Reply:
We could move the reading of `HOROVOD_GLOO_TIMEOUT_SECONDS` into this class.  It's never set to anything else in practice.

Reply:
There is no gloo:TimeoutException, that would have been too good... Then I would rather add the env var to the error message, though it is bad practise, but moving the env var in this code would be worse I think.

Reply:
I think having the env variable in the error is very useful.  Otherwise users will not know how to resolve the issue, and will ask questions on GitHub that could be answered by the error message (this has happened a few times).  I agree it's making an assumption that isn't guaranteed by the interface, but I think it's worth it.  What do you think?

Reply:
Yes, the information is more useful for users than clean code separation for developers in this case, I will go for it.

Reply:
I had to move the constants into `gloo_context.h`: https://github.com/horovod/horovod/pull/2436/commits/6cd9213d634397a9ac67c9472c66a4df4d6d4b3a